sklearn/model_selection/tests/test_successive_halving.py

```diff
@@ -0,0 +1,403 @@
+import pytest
+from scipy.stats import norm, randint
+import numpy as np
+
+from sklearn.datasets import make_classification
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import HalvingGridSearchCV
+from sklearn.model_selection import HalvingRandomSearchCV
+from sklearn.model_selection import KFold, ShuffleSplit
+from sklearn.model_selection._search_successive_halving import (
+    _SubsampleMetaSplitter, _top_k, _refit_callable)
+
+
+class FastClassifier(DummyClassifier):
+    """Dummy classifier that accepts parameters a, b, ... z.
+
+    These parameter don't affect the predictions and are useful for fast
+    grid searching."""
+
+    def __init__(self, strategy='stratified', random_state=None,
+                 constant=None, **kwargs):
+        super().__init__(strategy=strategy, random_state=random_state,
+                         constant=constant)
+
+    def get_params(self, deep=False):
+        params = super().get_params(deep=deep)
+        for char in range(ord('a'), ord('z') + 1):
+            params[chr(char)] = 'whatever'
+        return params
+
+
+@pytest.mark.parametrize('Est', (HalvingGridSearchCV, HalvingRandomSearchCV))
+@pytest.mark.parametrize(
+    ('aggressive_elimination,'
+     'max_resources,'
+     'expected_n_iterations,'
+     'expected_n_required_iterations,'
+     'expected_n_possible_iterations,'
+     'expected_n_remaining_candidates,'
+     'expected_n_resources,'), [
+         # notice how it loops at the beginning
+         (True, 'limited', 4, 4, 3, 1, [20, 20, 60, 180]),
+         # no aggressive elimination: we end up with less iterations and more
+         # candidates at the end
+         (False, 'limited', 3, 4, 3, 3, [20, 60, 180]),
```

Comment:
Aggressive elimination is described with:

> This is only relevant in cases where there isn't enough resources to reduce the candidates to at most `ratio` in the last iteration.

but here, even with aggressive_elimination=False the number of candidates is reduced to `ratio`.

Why is it useful for us to land up with a single candidate in the final iteration? Would we not rather try 3 candidates in the last round?

Reply:
> even with aggressive_elimination=False the number of candidates is reduced to ratio.

`n_remaining_candidates` indicates the number of candidates that remain **after** the last iteration. Those (3) remaining candidates weren't evaluated, and as you noted, we should ideally be evaluating (at most) 3 candidates in the last round.

I have clarified the docstrings and also included `n_candidates_`. I hope it's now clearer that with `aggressive_elemination=False`, the last iteration actually evaluates 7 candidates.